core/model/src/main/java/org/eclipse/rdf4j/model/impl/GenericSimpleStatement.java

```diff
@@ -0,0 +1,110 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ ******************************************************************************/
+
+package org.eclipse.rdf4j.model.impl;
+
+import java.util.Objects;
+
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Resource;
+import org.eclipse.rdf4j.model.Statement;
+import org.eclipse.rdf4j.model.Value;
+import org.eclipse.rdf4j.model.base.AbstractStatement;
+
+/**
+ * A simple default implementation of the {@link Statement} interface for statements that don't have an associated
+ * context. For statements that do have an associated context, {@link ContextStatement} can be used.
+ *
+ * @see SimpleValueFactory
+ */
+public class GenericSimpleStatement<R extends Resource, I extends IRI, V extends Value> extends AbstractStatement {
```

Comment:
Why do we need a new class `GenericSimpleStatement` and `GenericContextStatement`? What is their relation to the already-existing `SimpleStatement` and `ContextStatement`?

Reply:
The use java generics, which gives a decent performance boost by allowing the JVM to skip type checking and casting.

Reply:
They are otherwise copies of SimpleStatement and ContextStatement.

Reply:
I could make SimpleStatement and ContextStatement extend these two new classes, that would mean that we are pretty much moving all the code from those classes to the two new generic variants.

Reply:
What I don't fully understand is why these classes are in the model module, if the performance improvements are specifically targeted at the memory store. I'm sorry, I'm a bit short on time to look more closely, so don't take this as dismissing the idea, I'm just trying to get my head around it, with limited time.

Reply:
It's because `MemStatement` extends `ContextStatement`. And alternative solution would be to have `MemStatement` implement `Statement` directly. The reason the `MemStatement` needs to be able to specify what classes are used for s, p, o and c is because the nature of the MemoryStore allows us to use "==" to check if two MemIRI (and other Mem****) are the same without ever needing to use `.equals(...)`. Performance wise there is a decent benefit in being able to specify that MemStatement uses MemResource, MemIRI and MemValue. 

Adding GenericSimpleStatement and GenericContextStatement to the model module makes sense because this is where SimpleStatement and ContextStatement reside. It might make even more sense if I made them extend each other and deprecated the old ones.

e.g.

```
@Deprecated
public class SimpleStatement extends GenericSimpleStatement<Resource, IRI, Value> {
...
}

@Deprecated
public class ContextStatement extends GenericContextStatement<Resource, IRI, Value, Resource> {
...
}
```

We could also just skip having both a GenericSimpleStatement and a GenericContextStatement. Instead we could just have GenericStatement which implements the full Statement interface. 

I am hoping that the generic versions can be used in other sails too. And I am also hoping to introduce more generics in general, particularly at the Sail level since that code isn't user facing.

Reply:
Fwiw I never really liked the separation of SimpleStatement and ContextStatement in the first place - IMO it's a leftover from when "contexts" / "quads" / "named graphs" were still an experimental thing.

Reply:
I'll go with GenericStatement then.

Reply:
And I can mark it internal use only, for the time being.